components/bigqueryDS.js

```diff
@@ -0,0 +1,139 @@
+  }
+}
+
+function sanitizeTimestamp(str) {
+  // new Date("2022-04-14 09:24:34") generates  'Invalid Date' on safari.  Node seems to be happy ... so fix it up
+  const date = new Date(str);
+  return date.toISOString();
+}
+
+function normalize(entry) {
+  const normalizedEntry = {
+    insertId: entry.insertId,
+    logName: entry.logName,
+    receiveTimestamp: sanitizeTimestamp(entry.receiveTimestamp),
+    timestamp: sanitizeTimestamp(entry.timestamp),
+    jsonPayload: {
+      errorResponse: entry.jsonPayload_errorResponse,
+      request: safeJSONParse(entry.jsonPayload_request),
+      response: safeJSONParse(entry.jsonPayload_response),
+      "@type": entry.jsonPayload_type,
+    },
+    labels: safeJSONParse(entry.labels),
+    resource: {
+      labels: safeJSONParse(entry.resource_labels),
+    },
+  };
+  return normalizedEntry;
+}
+
+class Bigquery extends Datasource {
+  constructor(argv) {
+    super(argv);
+    this.dataset = argv.bigquery;
+    //this.timerangeClause = '
+  }
+
+  async bq(query) {
+    const cmd = `bq query --max_rows=1000 --format=json --nouse_legacy_sql '${query}'`;
+    console.log("Running bq command", cmd);
+    const ret = await exec(cmd, { maxBuffer: 1024 * 1024 * 30 });
+    //console.log("gots output", ret);
+    return safeJSONParse(ret.stdout);
+  }
+  /*
+   * Extract trip_ids from vehicle logs, and query again to
+   * get the createTrip/update trip calls that may not be labeled
+   * with a vehicle
+   */
+  async fetchTripLogsForVehicle(vehicle_id, vehicleLogs) {
+    console.log("Loading trip logs for vehicle id", vehicle_id);
+    const trip_ids = _(vehicleLogs)
+      .map((x) => _.split(_.get(x, "labels.trip_id"), ","))
+      .flatten()
+      .uniq()
+      .compact()
+      .value();
+    const tripPromises = trip_ids.map(async (trip_id) => {
+      const createLogs = await this.bq(
+        `select * from \`${this.dataset}.create_trip\` where json_value(labels.trip_id)="${trip_id}" order by timestamp desc limit 10`
+      );
+      const updateLogs = await this.bq(
+        `select * from \`${this.dataset}.update_trip\` where json_value(labels.trip_id)="${trip_id}" order by timestamp desc limit 10`
+      );
+      return [createLogs, updateLogs];
+    });
+
+    const tripLogs = _(await Promise.all(tripPromises))
+      .flatten()
+      .flatten()
+      .map(normalize)
+      .value();
+
+    return tripLogs;
+  }
+
+  /*
+   * Extract task_ids from vehicle logs, and query again to
+   * get the createTask/updateTask calls that may not be labeled
+   * with a vehicle
+   */
+  async fetchTaskLogsForVehicle(vehicle_id, vehicleLogs) {
+    console.log("checking for", vehicle_id, vehicleLogs);
+    return [];
+  }
+
+  async fetchVehicleLogs(vehicle_id, trip_id) {
+    let whereClause;
+    if (vehicle_id) {
+      whereClause = `json_value(labels.vehicle_id)="${vehicle_id}"`;
+    } else {
+      whereClause = `json_value(labels.trip_id) LIKE "%${trip_id}%"`;
+    }
```

Comment:
Consider being more defensive w/ https://cloud.google.com/bigquery/docs/parameterized-queries#node.js 

```suggestion
    let whereClause, params = {};
    if (vehicle_id) {
      whereClause = `json_value(labels.vehicle_id)=@vehicleId`;
      params['vehicleId'] = `${vehicle_id}`;
    } else {
      whereClause = `json_value(labels.trip_id) LIKE @tripId`;
      params['tripId'] = `%${trip_id}%`;
    }
```

Reply:
switched to using biquery node library.  Much cleaner